File: packages/icoder-python/icoder_sdk/resources/facts.py
"""Facts extraction resource."""

from typing import Optional

from ..client import iCoDerClient
from ..request_options import RequestOptions
from ..types import FactExtractResponse, FactItem, FactUsageInfo


class FactsResource:
    def __init__(self, client: iCoDerClient):
        self._client = client

    def extract(
        self,
        text: str,
        output_language: str = "zh-CN",
        request_options: Optional[RequestOptions] = None,
    ) -> FactExtractResponse:
        if not text.strip():
            raise ValueError("text must not be empty")
        resp = self._client.post("/api/v2/tools/extract-facts", json={
            "context": [{"type": "text", "text": text}],
            "outputLanguage": output_language,
        }, request_options=request_options)
        resp.raise_for_status()
        data = resp.json()
        return FactExtractResponse(
            facts=[FactItem(
                group=str(item.get("group", "")),
                text=str(item.get("text", "")),
                value=str(item.get("value", "")),
            ) for item in data.get("facts", [])],
            output_language=str(data.get("outputLanguage", output_language)),
            usage_info=FactUsageInfo(
                credits_consumed=float(data.get("usageInfo", {}).get("creditsConsumed", 0)),
            ),
        )

File: packages/icoder-sdk/src/resources/facts.ts
// Facts resource
import type { AxiosInstance } from 'axios';
import type { FactExtractRequest, FactExtractResponse } from '../types.js';
import { requestConfig, type iCoDerRequestOptions } from '../request-options.js';

export class FactsResource {
  constructor(private http: AxiosInstance) {}

  /** Extract structured clinical facts with the Corti-compatible v2 contract. */
  async extract(
    request: FactExtractRequest,
    requestOptions?: iCoDerRequestOptions,
  ): Promise<FactExtractResponse>;
  async extract(
    text: string,
    outputLanguage?: string,
    requestOptions?: iCoDerRequestOptions,
  ): Promise<FactExtractResponse>;
  async extract(
    requestOrText: FactExtractRequest | string,
    outputLanguageOrOptions: string | iCoDerRequestOptions = 'zh-CN',
    requestOptions?: iCoDerRequestOptions,
  ): Promise<FactExtractResponse> {
    const outputLanguage = typeof outputLanguageOrOptions === 'string'
      ? outputLanguageOrOptions
      : 'zh-CN';
    const options = typeof requestOrText === 'string'
      ? requestOptions
      : typeof outputLanguageOrOptions === 'string'
        ? undefined
        : outputLanguageOrOptions;
    const request: FactExtractRequest = typeof requestOrText === 'string'
      ? { context: [{ type: 'text', text: requestOrText }], outputLanguage }
      : requestOrText;
    if (!request.context.some((item) => item.text.trim())) {
      throw new Error('context must contain at least one non-empty text item');
    }
    const { data } = await this.http.post<FactExtractResponse>('/api/v2/tools/extract-facts', {
      ...request,
      outputLanguage: request.outputLanguage || 'zh-CN',
    }, requestConfig(options));
    return data;
  }
}
